src/internal/operators/timeout.ts

```diff
@@ -1,90 +1,395 @@
-import { async } from '../scheduler/async';
-import { TimeoutError } from '../util/TimeoutError';
-import { MonoTypeOperatorFunction, SchedulerLike } from '../types';
-import { timeoutWith } from './timeoutWith';
-import { throwError } from '../observable/throwError';
+/** @prettier */
+import { asyncScheduler } from '../scheduler/async';
+import { MonoTypeOperatorFunction, SchedulerLike, OperatorFunction, ObservableInput, TeardownLogic } from '../types';
 import { defer } from '../observable/defer';
+import { isValidDate } from '../util/isDate';
+import { OuterSubscriber } from '../OuterSubscriber';
+import { Subscriber } from '../Subscriber';
+import { Subscription } from '../Subscription';
+import { lift } from '../util/lift';
+import { Operator } from '../Operator';
+import { Observable } from '../Observable';
+import { subscribeToResult } from '../util/subscribeToResult';
 
-function timeoutErrFactory() { return throwError(new TimeoutError()); }
+export interface TimeoutConfig<T, R> {
+  /**
+   * The time allowed between values from the source before timeout is triggered.
+   */
+  each?: number;
+
+  /**
+   * The relative time as a `number` in milliseconds, or a specific time as a `Date` object,
+   * by which the first value must arrive from the source before timeout is triggered.
+   */
+  first?: number | Date;
+
+  /**
+   * The scheduler to use with time-related operations within this operator. Defaults to {@link asyncScheduler}
+   */
+  scheduler?: SchedulerLike;
+
+  /**
+   * A factory used to create observable to switch to when timeout occurs. Provides
+   * some information about the source observable's emissions and what delay or
+   * exact time triggered the timeout.
+   */
+  with?: (info: TimeoutInfo<T>) => ObservableInput<R>;
+
+  /**
+   * Optional additional metadata you can provide to code that handles
+   * the timeout, will be provided through the {@link TimeoutError}.
+   * This can be used to help identify the source of a timeout or pass along
+   * other information related to the timeout.
+   */
+  meta?: any;
+}
+
+export interface TimeoutInfo<T> {
+  /** Optional metadata that was provided to the timeout configuration. */
+  meta: any;
+  /** The number of messages seen before the timeout */
+  seen: number;
+  /** The last message seen */
+  lastValue: T | null;
```

Comment:
This might sound odd, but this wasn't an arbitrary choice. What's weird is I can't find a precedent here for JavaScript proper, all of those error types don't really have additional properties that I can find, but for DOM errors, properties are usually initialized with `null`, `0`, or `""`, depending on what is supposed to be in there. Consider `new ErrorEvent(new Error())`... `target` is `null`.

I suspect this is a "same shapedness" optimization thing, but It feels to me like the property should be defined (so runtime tests could confirm the type) and that it should initialize to `null`, just to be in line with what is done in the DOM, etc. Should actually minify a little better, too.

I'm willing to accept arguments either way though. Not married to it. Just telling you my rationale.

Reply:
My way of looking at it has less to do with what's in the `Error` - or what's in DOM errors - and more to do with the way the language works, these days.

Some time ago, I made the decision to ban `null` wherever possible in my code, but especially as a representation for 'no value'. The problem is that `null` just doesn't play nice with modern JS idioms:

```js
function f(a = 42) {
    console.log(a);
}
f(undefined); // 42
f(null); // null
```

```js
const { a = 42, b = 54 } = { a: undefined, b: null };
console.log(a); // 42
console.log(b); // null
```

```js
const a = null?.prop;
console.log(a); // undefined
console.log(null && null.prop); // null
```

Reply:
Either way... I definitely don't think it should be `lastValue?`.. as the property should at least be defined on the object, IMO.